From: George Beshers <gbeshers@redhat.com>
Date: Fri, 18 Dec 2009 22:26:55 -0500
Subject: [uv] x86: xpc_make_first_contact hang due to not accepting ACTIVE state
Message-id: <20091218171313.23145.1726.sendpatchset@localhost.localdomain>
Patchwork-id: 22198
O-Subject: [RHEL6 PATCH 4/6] x86: uv: xpc_make_first_contact hang due to not
	accepting ACTIVE state
Bugzilla: 546695
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

commit dbd2918ec65c35f36bb102c88eafe87be0552f6f
Author: Robin Holt <holt@sgi.com>
Date:   Tue Dec 15 16:47:58 2009 -0800

    X86: uv: xpc_make_first_contact hang due to not accepting ACTIVE state

    Many times while the initial connection is being made, the contacted
    partition will send back both the ACTIVATING and the ACTIVE
    remote_act_state changes in very close succescion.  The 1/4 second delay
    in the make first contact loop is large enough to nearly always miss the
    ACTIVATING state change.

    Since either state indicates the remote partition has acknowledged our
    state change, accept either.

    Signed-off-by: Robin Holt <holt@sgi.com>
    Cc: Jack Steiner <steiner@sgi.com>
    Cc: Ingo Molnar <mingo@elte.hu>
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

Signed-off-by: Aristeu Rozanski <arozansk@redhat.com>

diff --git a/drivers/misc/sgi-xp/xpc_uv.c b/drivers/misc/sgi-xp/xpc_uv.c
index 19bd7b0..241ea5f 100644
--- a/drivers/misc/sgi-xp/xpc_uv.c
+++ b/drivers/misc/sgi-xp/xpc_uv.c
@@ -1024,7 +1024,8 @@ xpc_make_first_contact_uv(struct xpc_partition *part)
 	xpc_send_activate_IRQ_part_uv(part, &msg, sizeof(msg),
 				      XPC_ACTIVATE_MQ_MSG_SYNC_ACT_STATE_UV);
 
-	while (part->sn.uv.remote_act_state != XPC_P_AS_ACTIVATING) {
+	while (!((part->sn.uv.remote_act_state == XPC_P_AS_ACTIVATING) ||
+		 (part->sn.uv.remote_act_state == XPC_P_AS_ACTIVE))) {
 
 		dev_dbg(xpc_part, "waiting to make first contact with "
 			"partition %d\n", XPC_PARTID(part));
